// realm-object-server/dist/realms/UserPermissionRealm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RealmFactory_1 = require("../RealmFactory");
const AdminRealm_1 = require("./AdminRealm");
class UserSpecificPermission {
}
UserSpecificPermission.schema = {
    name: 'Permission',
    properties: {
        userId: { type: 'string', indexed: true },
        path: { type: 'string', indexed: true },
        mayRead: { type: 'bool', optional: false },
        mayWrite: { type: 'bool', optional: false },
        mayManage: { type: 'bool', optional: false },
        updatedAt: { type: 'date', optional: false },
    }
};
exports.UserSpecificPermission = UserSpecificPermission;
function CreateLegacyUserPermissionRealmDefinition(userId) {
    return {
        remotePath: `/${userId}/__permission`,
        schema: [
            RealmFactory_1.createRealmSchema(UserSpecificPermission)
        ],
    };
}
exports.CreateLegacyUserPermissionRealmDefinition = CreateLegacyUserPermissionRealmDefinition;
function CreateUserPermissionRealmDefinition(userId) {
    return {
        remotePath: `/${userId}/__perm`,
        schema: [
            RealmFactory_1.createRealmSchema(AdminRealm_1.UniquePermission),
        ],
    };
}
exports.CreateUserPermissionRealmDefinition = CreateUserPermissionRealmDefinition;
