// realm-object-server/dist/realms/WildcardPermissionRealm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RealmFactory_1 = require("../RealmFactory");
const AdminRealm_1 = require("./AdminRealm");
class WildcardSpecificPermission {
}
WildcardSpecificPermission.schema = {
    name: 'Permission',
    properties: {
        userId: { type: 'string', default: '*' },
        path: { type: 'string', indexed: true },
        mayRead: { type: 'bool', optional: false },
        mayWrite: { type: 'bool', optional: false },
        mayManage: { type: 'bool', optional: false },
        updatedAt: { type: 'date', optional: false },
    }
};
exports.WildcardSpecificPermission = WildcardSpecificPermission;
exports.LegacyWildcardPermissionRealm = {
    remotePath: `/__wildcardpermissions`,
    schema: [
        RealmFactory_1.createRealmSchema(WildcardSpecificPermission)
    ],
};
exports.WildcardPermissionRealm = {
    remotePath: `/__perm`,
    schema: [
        RealmFactory_1.createRealmSchema(AdminRealm_1.UniquePermission),
    ],
};
