// realm-object-server/dist/realms/DefaultRealm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RealmFactory_1 = require("../RealmFactory");
class DefaultRealmVersion {
}
DefaultRealmVersion.schema = {
    name: '__DefaultRealmVersion',
    primaryKey: 'id',
    properties: {
        id: { type: 'int', optional: false },
        version: { type: 'int', optional: false },
    }
};
DefaultRealmVersion.CurrentVersion = 1;
DefaultRealmVersion.SingletonId = 0;
exports.DefaultRealmVersion = DefaultRealmVersion;
function initialize(realm) {
    realm.beginTransaction();
    const single = realm.objectForPrimaryKey(DefaultRealmVersion.schema.name, DefaultRealmVersion.SingletonId);
    const should_update = !single || single.version !== DefaultRealmVersion.CurrentVersion;
    if (should_update) {
        realm.create(DefaultRealmVersion.schema.name, {
            id: DefaultRealmVersion.SingletonId,
            version: DefaultRealmVersion.CurrentVersion
        }, true);
        realm.commitTransaction();
    }
    else {
        realm.cancelTransaction();
    }
}
exports.initialize = initialize;
exports.Definition = {
    remotePath: '/default',
    localPath: 'realms/default.realm',
    syncLabel: 'default',
    schema: [
        RealmFactory_1.createRealmSchema(DefaultRealmVersion),
    ],
};
